app/javascript/controllers/replace_controller.js

```diff
@@ -0,0 +1,24 @@
+import { Controller } from "@hotwired/stimulus"
+import { createConsumer } from "@rails/actioncable"
+
+// Connects to data-controller="post-replace"
+export default class extends Controller {
+  static values = { postId: Number }
+  static targets = ["post"]
+
+  // When connected, this will open a connection with the PostChannel using the id from the values
+  // The received method indicates what happens with the data received.
+  // In this example it is a one line replacement but can be a more complicated method
+  connect() {
+    this.channel = createConsumer().subscriptions.create(
+      { channel: "PostChannel", id: this.postIdValue },
+      { received: data => this.postTarget.outerHTML = data }
```

Comment:
the `data => this.postTarget.outHTML = data` should be sufficient to replace a whole partial in place. 

If you do not want to refactor a whole page, you could use the starting example I gave in class where you have a new partial `_script_content.html.erb` that is super simple: 

```html 
<div data-controller="replace", data-replace-script-id-value="<%= script.id %>">
  <p data-replace-script-target="script">
      <%= script.script_body %>
  </p>
</div>
```
Then from the /show.html.erb where the script_body already is shown, you do a render of this partial:
`<%= render partial "scripts/script_content", locals: { script: @script } %>`

This prevents needing to refactor everything inside the replace controller. One downside is that this may require re-thinking and updating how the Script edit feature works since it writes the html and it may need to write this as well ( not sure without further investigation ). 

Thus the flow would be this: 
1. The View has a button "update/regenerate". User clicks
2. Rails controller ( def update ) gets the call and updates the script details ( but not script_body ) 
3. Rails controller calls a background job ( "StartSriptGenerationJob" )
4. ScriptGenerationJob is called with the ID of the Script which it then uses to find the script.
5. ScriptGeneration job then does `script.generate_script` which is a method in the Script model.
6.`generate_script` then takes the details from the script and calls the OpenAI service you have. It gets back thew new details. 
7. `generate_script` then takes the new content from the AI and does `self.update(script_body: content)`. 
8. it then calls a `broadcast` method which creates a new partial and broadcasts that to the ScriptChannel. 
9. The stimulus controller listens and replaces the old content with the new content. 

At this point you can stop and have a working feature.
The good: 
1. It is simple and thus quick to implement
2. The background job of step 3 means the user can interact with the page while Open AI is called by the server. 
The bad: 
1. It feels weird for the User to click a button and nothing immediately happens ( not even a loading icon )
2. It feels were for the text to be 100% changed from one second to the next. Especially if the text isn't super easy to stop the difference on


How can you address the bad? 
1. In the ScriptController or as the first action of the `generate_script`, you can broadcast a different partial. This partial, "views/scripts/_pending.html.erb" can be connected to a loading_controller.js that gives the users some visuals while you wait. This new partial would still need to connect to the `replace_controller.js` and provide it a target. Then when the AI produces the script it continues to step 7 above like normal and everything gets updated. 
2. stream the output bit by bit. This would be using Yann's example. 

The benefit of option 1 is that it is just a minor tweek to the simplest version of this while providing immediate user feedback. The bad is that it can be frustrating for a user to wait and not know the progress or see results. Also if they "refresh" the page they won't see any progrss. 

The benefit of option 2 is that the user gets drip fed info as soon as it becomes available. This can also be combined with option 1 to provide absolutely immediate feedback and then very quick the dripfeed. It may have an oddity with the Refresh but perhaps not. It is, however, a bit more complicated. Since Yann has provided code, though, it should be okay to aim for. 

Either option I think will get you to where you want to go. So choose the one that meets your timeline requirements and requires the least refactoring of other code not related to this. If both are good, choose Yann's since it gives you a chance to experiment with more complex code.